components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -1287,6 +1330,67 @@ private class MonitoredItemWrapper {
                 return eventFilter;
             }
 
+            public void ProcessMonitoredItemNotification(SubscriptionNotificationModel message, MonitoredItemNotificationModel monitoredItemNotification) {
+                var pendingAlarmsOptions = EventTemplate?.PendingAlarms;
+                if (pendingAlarmsOptions?.IsEnabled == true && monitoredItemNotification.Value.GetValue(typeof(EncodeableDictionary)) is EncodeableDictionary values) {
+                    var conditionIdIndex = pendingAlarmsOptions.ConditionIdIndex ?? -1;
+                    var retainIndex = pendingAlarmsOptions.RetainIndex ?? -1;
+                    if (conditionIdIndex != -1) {
+                        var conditionId = values[conditionIdIndex].Value.ToString();
+                        if (!PendingAlarmEvents.ContainsKey(conditionId) || retainIndex == -1) {
+                            PendingAlarmEvents[conditionId] = monitoredItemNotification;
+                            pendingAlarmsOptions.Dirty = true;
+                        }
+                        else {
+                            var existingValues = PendingAlarmEvents[conditionId].Value.GetValue(typeof(EncodeableDictionary)) as EncodeableDictionary;
+                            if (existingValues[retainIndex] != values[retainIndex]) {
+                                PendingAlarmEvents[conditionId] = monitoredItemNotification;
+                                pendingAlarmsOptions.Dirty = true;
+                            }
+                        }
+                    }
+                }
+                else {
+                    message.Notifications?.Add(monitoredItemNotification);
+                }
+            }
+
+            private void SendPendingAlarms(bool snapshot) {
+                var pendingAlarmsOptions = EventTemplate?.PendingAlarms;
+                if (pendingAlarmsOptions == null) {
+                    return;
+                }
+
+
+                if (pendingAlarmsOptions.IsEnabled == true &&
+                    pendingAlarmsOptions.RetainIndex.HasValue == true &&
+                    (snapshot || pendingAlarmsOptions.Dirty == true)) {
+                    var retainIndex = EventTemplate.PendingAlarms.RetainIndex.Value;
+                    var message = new SubscriptionNotificationModel {
+                        ServiceMessageContext = Item.Subscription?.Session?.MessageContext,
+                        ApplicationUri = Item.Subscription?.Session?.Endpoint?.Server?.ApplicationUri,
+                        EndpointUrl = Item.Subscription?.Session?.Endpoint?.EndpointUrl,
+                        SubscriptionId = (Item.Subscription?.Handle as SubscriptionWrapper)?.Id,
+                        Timestamp = DateTime.UtcNow,
+                        Notifications = PendingAlarmEvents.Values
+                            .Where(x => !snapshot || (x.Value.GetValue(typeof(EncodeableDictionary)) as EncodeableDictionary)[retainIndex].Value.GetValue<bool>(false) == true)
```

Comment:
The where clause makes me think... We shouldn't need it. If we simply **remove** any alarm with retain unset from the cache. Then the entire cache can always be sent (because it only has alarm with retain set). And no filtering needed here. We can discuss if unclear.

Reply:
Yes, as stated above we need to discuss this algorithm and find out what's the best use case.

Reply:
UPDATE: Discussed with the customer and code is updated. Please revisit the PR